src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,257 @@
+    internal sealed class CSharpConvertLocalFunctionToMethodCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+            if (!node.IsKind(SyntaxKind.LocalFunctionStatement, out LocalFunctionStatementSyntax localFunction))
+            {
+                return;
+            }
+
+            if (localFunction.ContainsDiagnostics)
+            {
+                return;
+            }
+
+            if (!localFunction.Identifier.Span.IntersectsWith(context.Span))
+            {
+                return;
+            }
+
+            if (!localFunction.Parent.IsKind(SyntaxKind.Block, out BlockSyntax parentBlock))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)));
+        }
+
+        private static async Task<Document> UpdateDocumentAsync(
+            SyntaxNode root,
+            Document document,
+            BlockSyntax parentBlock,
+            LocalFunctionStatementSyntax localFunction,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declaredSymbol = semanticModel.GetDeclaredSymbol(localFunction, cancellationToken);
+
+            var dataFlow = semanticModel.AnalyzeDataFlow(localFunction.Body);
+
+            var isStatic =
+                declaredSymbol.ContainingType.IsStatic ||
+                declaredSymbol.ContainingSymbol.IsStatic ||
+                dataFlow.Captured.All(capture => !capture.IsThisParameter());
+
+            var capturedTypeParameters = GetCapturedTypeParameters(declaredSymbol);
+            var unusedTypeParameters = capturedTypeParameters.ToList();
+            foreach (var id in localFunction.DescendantNodes().OfType<IdentifierNameSyntax>())
+            {
+                var sym = semanticModel.GetSymbolInfo(id).Symbol;
+                unusedTypeParameters.Remove(sym?.OriginalDefinition);
+            }
+
+            var typeParameters = capturedTypeParameters
+                .Except(unusedTypeParameters)
+                .Reverse()
+                .Cast<ITypeParameterSymbol>()
+                .ToList();
+
+            var captures = dataFlow.Captured.OfType<ILocalSymbol>();
+            var capturesAsParameters = captures.Select(capture => Parameter(
```

Comment:
oh man, now i really don't like the 'using static'.  it looks like you're calling a local method.  i got very confused.

Reply:
Note: you're generating syntax here.  My suggestion would be to generate actual Symbols.  You can use CodeGenerationSymbolFactory for this.